### DISCUSSION TITLE
NPM BUILD takes 40 minutes

### DISCUSSION THREAD

**Discussion by n1603** [scality/metalk8s#3068] [General]
Hi,
I have seen multiple times NPM BUILD takes more than 30 minutes to finish. Could you please suggest if this is expected behaviour or its some issue specific to my setup. I have consistently observed this behaviour multiple times.

**Comment by ChengYanJin** [comment]
Hello @n1603 ,

Normally it takes us around 1 min to build the UI on our dev environment. We never encounter a huge building time.

Would you confirm the CPU is not overloaded?
Any chance to know the version of npm and node on your environment? 

It would be great if you can share with us more details!

**Reply by n1603** [reply]
@ChengYanJin,
Thanks for looking in this. Below are the details.

a. There is no issues related to CPU, the VM is only doing build and has 8 cpus allocated. load avg was fine on vm.

b. It gets stuck at below stage while build : 

.  NPM BUILD      _build/ui/index.html

c. Below is the docker ps command output and log

> docker ps -a | grep metalk8s-ui-builder
d51109e04092   metalk8s-ui-builder:latest    "/entrypoint.sh /ent…"   **43 minutes ago   Up 43 minutes**                       priceless_brahmagupta`

> node         1     0  0 03:34 ?        00:00:00 /bin/bash /entrypoint.sh build
node       460     1  0 04:17 ?        00:00:00 npm
node       471   460  0 04:17 ?        00:00:00 sh -c react-app-rewired build
node       472   471  0 04:17 ?        00:00:00 node /home/node/node_modules/.bin/react-app-rewired build
node       479   472 99 04:17 ?        00:01:42 node /home/node/node_modules/react-app-rewired/scripts/build.js
node       490   479 19 04:18 ?        00:00:02 /usr/local/bin/node /home/node/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/workers/processChild.js
node       497   479  2 04:18 ?        00:00:00 /usr/local/bin/node /home/node/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/workers/processChild.js
node       504   479 11 04:18 ?        00:00:01 /usr/local/bin/node /home/node/node_modules/terser-webpack-plugin/node_modules/jest-worker/build/workers/processChild.js
node       511     0  1 04:18 pts/0    00:00:00 /bin/sh


d. npm version 
> $ npm version (from container)
{ 'metalk8s-platform-ui': '0.1.0',
  npm: '6.9.0',
  ares: '1.15.0',
  brotli: '1.0.7',
  cldr: '35.1',
  http_parser: '2.8.0',
  icu: '64.2',
  modules: '64',
  napi: '4',
  nghttp2: '1.34.0',
  node: '10.16.0',
  openssl: '1.1.1b',
  tz: '2019a',
  unicode: '12.1',
  uv: '1.28.0',
  v8: '6.8.275.32-node.52',
  zlib: '1.2.11' }

**Reply by ChengYanJin** [reply]
Hi @n1603, thanks for the information.
Could you try to run `npm run build` in ui/ directory to see if you have the same slowness issue?

**Reply by n1603** [reply]
Hi @ChengYanJin, Thanks, could you please suggest what all pre-requisites tools need to be installed. I tried this and got below error : 

>  npm run build
>
>  metalk8s-platform-ui@0.1.0 build
>  react-app-rewired build

>sh: react-app-rewired: command not found
>npm ERR! code 127
>npm ERR! path /Users/admin/metalk8s/ui
>npm ERR! command failed
>npm ERR! command sh -c react-app-rewired build

**Reply by NicolasT** [reply]
You'll need to run `npm install` before `npm run build`, isn't it?

**Reply by n1603** [reply]
Thanks for the details.  I see it takes around 3 minutes to finish "npm run build" when run directly under ui directory.

**Reply by ChengYanJin** [reply]
Hi @n1603, I would say around 3 minutes is as expected. 
Can you explicitly describe the steps you followed to build the project? And do you see any interesting logs related to NPM BUILD that could share with us?

**Reply by n1603** [reply]
@ChengYanJin, Thanks for checking, I will debug this further, could not get back to it.

**Reply by n1603** [reply]
@ChengYanJin, I tried to run the build from inside of Docker container. I see below it gets stuck extracting : types/js-yaml@3.12.5 and other places for long time....


`
$ id
uid=1000(node) gid=1000(node) groups=1000(node)
$ npm ci
[                  ] \ extractTree: sill extract @types/js-yaml@3.12.5 extracted to /home/node/node_modules/@types/js-yaml (1028ms)

[                  ] \ extractTree: sill extract @types/js-yaml@3.12.5 extracted to /home/node/node_modules/@types/js-yaml (1028ms)

[                  ] \ extractTree: sill extract @types/js-yaml@3.12.5 extracted to /home/node/node_modules/@types/js-yaml (1028ms)

[                  ] \ extractTree: sill extract @types/js-yaml@3.12.5 extracted to /home/node/node_modules/@types/js-yaml (1028ms)

...
[                  ] \ extractTree: sill extract symbol-observable@1.0.1 extracted to /home/node/node_modules/listr/node_modules/symbol-observable (7ms)
...
> cypress@3.8.3 postinstall /home/node/node_modules/cypress
> node index.js --exec install

Installing Cypress (version: 3.8.3)

 ✔  Downloaded Cypress
 ✔  Unzipped Cypress
 ✔  Finished Installation /home/node/.cache/Cypress/3.8.3

You can now open Cypress by running: node_modules/.bin/cypress open

https://on.cypress.io/installing-cypress

added 2213 packages in 2586.723s
`
added 2213 packages in **2586.723s**

**Comment by NicolasT** [comment]
@n1603 To better understand how people in the community use MetalK8s, would you mind chiming in at https://github.com/scality/metalk8s/discussions/3085 to provide some intel about your use-case? Thanks!
